go/vt/vtgate/gateway/discoverygateway.go

```diff
@@ -358,6 +378,33 @@ func (dg *discoveryGateway) withRetry(ctx context.Context, target *querypb.Targe
 	invalidTablets := make(map[string]bool)
 
 	for i := 0; i < dg.retryCount+1; i++ {
+		// Check if we should buffer MASTER queries which failed due to an ongoing
+		// failover.
+		// Note: We only buffer "!inTransaction" queries i.e.
+		// a) no transaction is necessary (e.g. critical reads) or
+		// b) no transaction was created yet.
+		if !inTransaction && target.TabletType == topodatapb.TabletType_MASTER {
+			// The next call blocks if we should buffer during a failover.
+			if retryDone, bufferErr := dg.buffer.WaitForFailoverEnd(ctx, target.Keyspace, target.Shard, err); bufferErr == nil {
```

Comment:
Let's test for bufferErr != nil here:

if retryDone, bufferErr := dg.buffer.WaitForFailoverEnd(ctx, target.Keyspace, target.Shard, err); bufferErr != nil {
  ...
  break
}

... no need to be in brackets here.